include/openthread/srp_server.h

```diff
@@ -0,0 +1,212 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *  This file defines the API for server of the Service Registration Protocol (SRP).
+ */
+
+#ifndef OPENTHREAD_SRP_SERVER_H_
+#define OPENTHREAD_SRP_SERVER_H_
+
+#include <stdint.h>
+
+#include <openthread/instance.h>
+#include <openthread/ip6.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup api-srp
+ *
+ * @brief
+ *  This module includes functions to control service registrations.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This structure represents a SRP service host.
+ *
+ */
+typedef void otSrpServerHost;
+
+/**
+ * This structure represents a SRP service.
+ *
+ */
+struct otSrpServerService
+{
+    char *              mFullName;  ///< The full service name.
```

Comment:
I think the user is stopped from changing the service content by returning `const otSrpService *`.

I did some refactoring and this structure removed by providing getters (e.g. `otSrpServeverServiceGetFullName()`) just as we do for `otSrpServerHost`.

Reply:
A `const otSrpService *` would not protect against access to the pointed char buffer.

In ` const` structure, the member variables would be treated as `const`, so no one can change or assign new values to them. The `mFullName` would be treated as `char * const mFullName` (which means the `mFullName` pointer value itself cannot change but the buffer to which it points can be changed) and not `const char *const mFullName`.

Here is small test for this:
```
    struct Test
    {
        char *mName;
    };

    char name[2];
    char name2[2];
    Test test;
    const Test *tp = &test;

    test.mName = name;
    name[0] = 'A';
    name[1] = 0;

    printf("name:\"%s\"\n", tp->mName);

    tp->mName[0] = 'B';    // This works even though `tp` is `const Test *`
    printf("name:\"%s\"\n", tp->mName);

    // tp->mName = name2; // This errors "cannot assign to variable 'tp' with const-qualified type 'const Test *'"
```

Reply:
The new model with opaque `Service` and getter APIs look good. Thank. I think it may add a bit to code size but for server code not sure if it matters much.

Reply:
Thanks for verifying this, I was not aware of this behavior.